### PR TITLE
Allow a more complex customized `myshopify_domain`

### PR DESCRIPTION
**Pull request by nelsonwittwer**
### What this PR does

1. Fixes bug where using a custom `myshopify_domain` with several subdomains would falsely classify the redirect as a phishing attack. This change replaces the regex with `Addressable::URI` which has far more support for all the variants of parsing a URI in ruby.
2. Adds some logging we forgot with previous PRs

### Reviewer's guide to testing

- Should be able to customize `myshopify_domain` with a domain that includes several subdomains and install the app
- Should still be able to use default `myshopify.com` myshopify_domain and install the app

### Checklist

Before submitting the PR, please consider if any of the following are needed:

- [ ] Update `CHANGELOG.md` if the changes would impact users
- [ ] Update `README.md`, if appropriate.
- [ ] Update any relevant pages in `/docs`, if necessary
- [ ] For security fixes, the [Disclosure Policy](https://github.com/Shopify/shopify_app/blob/master/SECURITY.md#disclosure-policy) must be followed.

### CODE REVIEW

File: test/shopify_app/utils_test.rb [hunk 1]
```diff
@@ -34,8 +42,7 @@ class UtilsTest < ActiveSupport::TestCase
   end
 
   ["myshop.com", "myshopify.com", "shopify.com", "two words", "store.myshopify.com.evil.com",
-   "/foo/bar", "foo.myshopify.io.evil.ru", "%0a123.myshopify.io",
-   "foo.bar.myshopify.io",].each do |bad_url|
```
[the remark this answers is not in the document]
Comment (nelsonwittwer) [reply test/shopify_app/utils_test.rb:38]: I started with the standard `URI.parse`, yes but ran into friction as it doesn't provide a way to return just the `domain` that we were looking to compare against. As I wanted to get away from regexes to make this function more resilient to URL changes, using a community vetted library to extract the domain seemed like a more sustainable approach.

File: lib/shopify_app/utils.rb [hunk 2]
```diff
@@ -6,19 +6,29 @@ module Utils
       "shopify.com",
       "myshopify.io",
       "myshopify.com",
+      "spin.dev",
     ].freeze
 
     def self.sanitize_shop_domain(shop_domain)
       myshopify_domain = ShopifyApp.configuration.myshopify_domain
       name = shop_domain.to_s.downcase.strip
+      name = "https://" + name unless name.include?("http://") || name.include?("https://")
```
[the remark this answers is not in the document]
Comment (nelsonwittwer) [reply lib/shopify_app/utils.rb:15]: I think I actually tried this before; unfortunately, in order to parse the URI you need the scheme 🤦 .

```ruby
uri = URI.parse(name)
uri.host 
  => nil

uri = URI.parse("http://" + name)
uri.host 
  => "my-shop.myshopify.com"

uri = Addressable::URI.parse(name)
uri.host 
  => nil
uri = Addressable::URI.parse("http://" + name)
uri.host
  => "my-shop.myshopify.com"
```

I think our safest bet will be to check to see if it starts with "http" as you suggested

[the remark this answers is not in the document]
Comment (nelsonwittwer) [reply lib/shopify_app/utils.rb]: I'm feeling much better about the solution I've just pushed. I believe it encapsulates your intent with leaning into parsing the URI instead relying on finding characters. Let me know what you think!

[the remark this answers is not in the document]
Comment (francisbeaudoin) [reply lib/shopify_app/utils.rb]: That implementation looks good and I don't see accepting risk associated with accepting _invalid_ domains such as the encoded one.

**Review by klenotiw — COMMENTED** [review]
Getting this error when I try to install on this branch. So hold off on merging this until I can find the cause.
Screenshot 2023-01-03 at 2 04 22 PM

**Review by klenotiw — APPROVED** [review]

**Pull request merged**